crypto/sm2/sm2_pmeth.c

```diff
@@ -163,26 +153,21 @@ static int pkey_sm2_decrypt(EVP_PKEY_CTX *ctx,
 static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
 {
     SM2_PKEY_CTX *smctx = ctx->data;
-    EC_GROUP *group;
     uint8_t *tmp_id;
 
     switch (type) {
     case EVP_PKEY_CTRL_EC_PARAMGEN_CURVE_NID:
-        group = EC_GROUP_new_by_curve_name(p1);
-        if (group == NULL) {
+        /*
+         * This control could be removed, which would signal it being
+         * unsupported.  However, that means that when the caller uses
+         * the correct curve, it may interpret the unsupported signal
+         * as an error, so it's better to accept the control, check the
+         * value and return a corresponding value.
+         */
+        if (p1 != NID_sm2) {
```

Comment:
I am sorry for the very late response - Just got back from the Chinese lunar new year vacation.

I would like to make a note here explaining that why we just use `NID_sm2` and ignore other curves. Though the current `NID_sm2` is the only recommended curve but SM2 algorithms (signature/encryption) can actually take other curves as well - if I recall correctly there exist test vectors for SM2 using other curves in some SM2 specification documents.

For OpenSSL I don't think we need to provide the ability to set other curves other than NID_sm2 because the NID_sm2 curve is the only one being used in practice. And when we talk about SM2 in different standard organization (ISO/IETF/...), there is also one curve NID_sm2.

But I still think it's better to leave a comment here (and the `paramgen` function below as well) that OpenSSL implements only the NID_sm2 curve and provides no other capability of setting other customized curves so that OpenSSL's implementation can only be valid with the SM2 test vectors that use NID_sm2.

@levitte ,what do you think?

Reply:
I didn't limit it like that. What I did was to change the *default* behaviour, but you can still change the `EVP_PKEY` type with `EVP_PKEY_set_alias_type`